service/src/main/java/com/google/auto/service/processor/AutoServiceProcessor.java

```diff
@@ -243,23 +250,24 @@ private String getBinaryNameImpl(TypeElement element, String className) {
     return getBinaryNameImpl(typeElement, typeElement.getSimpleName() + "$" + className);
   }
 
-  private DeclaredType getProviderInterface(AnnotationMirror providerAnnotation) {
-
-    // The very simplest of way of doing this, is also unfortunately unworkable.
-    // We'd like to do:
-    //    ServiceProvider provider = e.getAnnotation(ServiceProvider.class);
-    //    Class<?> providerInterface = provider.value();
-    //
-    // but unfortunately we can't load the arbitrary class at annotation
-    // processing time. So, instead, we have to use the mirror to get at the
-    // value (much more painful).
-
-    Map<? extends ExecutableElement, ? extends AnnotationValue> valueIndex =
-        providerAnnotation.getElementValues();
-    log("annotation values: " + valueIndex);
-
-    AnnotationValue value = valueIndex.values().iterator().next();
-    return (DeclaredType) value.getValue();
+  /**
+   * Returns the contents of a {@code Class[]}-typed "value" field in a given {@code annotationMirror}.
+   *
+   * <p>This method is needed because directly reading the value of such a field from an
+   * AnnotationMirror throws: <pre>
+   * javax.lang.model.type.MirroredTypeException: Attempt to access Class object for TypeMirror Foo.
+   * </pre>
+   *
+   * @param annotationMirror The AnnotationMirror to read fields from, e.g. {@link AutoService}.
+   * @return a set of fully-qualified names of classes appearing in 'fieldName' on 'annotation' on
+   *     'element'.
+   */
+  private ImmutableSet<DeclaredType> getValueFieldOfClasses(AnnotationMirror annotationMirror) {
+    //noinspection unchecked
+    return ((List<AnnotationValue>) getAnnotationValue(annotationMirror, "value").getValue())
+            .stream()
+            .map(value -> MoreTypes.asDeclared((DeclaredType) value.getValue()))
```

Comment:
Either cast to `(TypeMirror)` and pass to `MoreTypes.asDeclared()` or just cast to `DeclaredType`.

Can you add this todo for me:

 `TODO(ronshapiro): class literals may not always be declared types, i.e. int.class, int[].class`

Reply:
I don't quite follow the first comment? Todo is here: 9f81d2a

Reply:
`MoreTypes.asDeclared()` is essentially a cast to a `DeclaredType`, so passing in something that's already declared should do nothing. so you can replace the line with this:
`.map(value -> (DeclaredType) value.getValue())`

Reply:
236f337